src/v/storage/types.h

```diff
@@ -118,8 +139,20 @@ class stm_manager {
         return result;
     }
 
+    // Ensures unresolved transactional messages are not seen by compaction.
+    model::offset max_compactible_offset() {
+        model::offset result = model::offset::max();
+        if (_comp_stm) {
+            result = _comp_stm->max_compactable_offset();
+        }
+        return result;
+    }
+
 private:
+    // Since compactable_stm is a snapshotable_stm, it will appear in this list
+    // as well as in the value for _comp_stm.
     std::vector<ss::shared_ptr<snapshotable_stm>> _stms;
+    ss::shared_ptr<compactable_stm> _comp_stm;
```

Comment:
do we need a special case for this stm type ?, would it be enough to use `max_collectible_offset` as a compaction boundary ?

Reply:
This was one of my first ideas. @rystsov indicated that this compaction codepath should be distinct from the snapshot/eviction code. We should chat about the definition of "max collectible" and all the use cases.

Reply:
IIUC, he's saying the value of having them separate is that it allows stm to control compaction and eviction separately, and asking if we think it is worthwhile.

Reply:
i would vote for using `max_collectible_offset` in compaction, there is no point in compacting a log if it is about to be delete either way, having a special case for compaction stm make the whole thing more complicated

Reply:
As much as I'd love to get this merged, I tend to agree with @mmaslankaprv. At least this is what I wanted to try first, because we can always add separate compaction control in the future if needed. I didn't see a need for this currently, but maybe in the future we introduce a new compaction algorithm or something that needs it?

Unless there are any objections, I'll whip up a new PR or add a patch which eliminates the separate compactable offset plumbing, so we can see the difference in approaches.